src/toast/ops/conviqt.py

```diff
@@ -0,0 +1,693 @@
+                if self.hwp_angle is not None:
+                    hwp_angle = views.shared[self.hwp_angle][view]
+                    psi_pol += 2 * hwp_angle
+                all_theta.append(theta)
+                all_phi.append(phi)
+                all_psi.append(psi)
+                all_psi_pol.append(psi_pol)
+
+        if len(all_theta) > 0:
+            all_theta = np.hstack(all_theta)
+            all_phi = np.hstack(all_phi)
+            all_psi = np.hstack(all_psi)
+            all_psi_pol = np.hstack(all_psi_pol)
+
+        if verbose:
+            timer.report_clear(f"compute pointing angles for detector {det}")
+        return all_theta, all_phi, all_psi, all_psi_pol
+
+    def get_buffer(self, theta, phi, psi, det, verbose):
+        """Pack the pointing into the conviqt pointing array"""
+        timer = Timer()
+        timer.start()
+        pnt = conviqt.Pointing(len(theta))
+        if pnt._nrow > 0:
+            arr = pnt.data()
+            arr[:, 0] = phi
+            arr[:, 1] = theta
+            arr[:, 2] = psi
+        if verbose:
+            timer.report_clear(f"pack input array for detector {det}")
+        return pnt
+
+    def convolve(self, sky, beam, detector, pnt, det, verbose):
+        timer = Timer()
+        timer.start()
+        convolver = conviqt.Convolver(
+            sky,
+            beam,
+            detector,
+            self.pol,
+            self.lmax,
+            self.beammmax,
+            self.order,
+            self.verbosity,
+            self.comm,
+        )
+        convolver.convolve(pnt)
+        if verbose:
+            timer.report_clear(f"convolve detector {det}")
+
+        # The pointer to the data will have changed during
+        # the convolution call ...
+
+        if pnt._nrow > 0:
+            arr = pnt.data()
+            convolved_data = arr[:, 3].astype(np.float64)
+        else:
+            convolved_data = None
+        if verbose:
+            timer.report_clear(f"extract convolved data for {det}")
+
+        del convolver
+
+        return convolved_data
+
+    def calibrate_signal(self, data, det, beam, convolved_data, verbose, detectors):
+        """By default, libConviqt results returns a signal that conforms to
+        TOD = (1 + epsilon) / 2 * intensity + (1 - epsilon) / 2 * polarization.
+
+        When calibrate = True, we rescale the TOD to
+        TOD = intensity + (1 - epsilon) / (1 + epsilon) * polarization
+        """
+        if not self.calibrate or beam.normalized():
+            return
+        timer = Timer()
+        timer.start()
+        offset = 0
+        for obs in data.obs:
+            if det not in obs.select_local_detectors(detectors):
+                continue
+            focalplane = obs.telescope.focalplane
+            epsilon = self._get_epsilon(focalplane, det)
+            # Make sure detector data output exists
+            obs.detdata.ensure(self.det_data, detectors=detectors)
```

Comment:
Since we called this before the outer detector loop, we don't need it here, where it would be called once per detector using the full set of detectors.

Reply:
Yes.